[Lit]: Add build time mangler for lit-html templates

fallaciousreasoning opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/44051

**Note:** ~~Needs Privacy/Security review~~ Approved!
https://github.com/brave/reviews/issues/1872

## Background

In Polymer, the HTML is compiled to a `HTMLTemplate` element which we're free to modify to our hearts desire. Lit makes use of `lit-html` to generate HTML from tagged template literals, which works super well if you don't need to modify someone else's template.

In `brave-core` we have a large number of existing HTML overrides for Polymer elements (especially in chrome://settings). Upstream is gradually migrating pages from Polymer to Lit, so we need a strategy to override these pages.

Unfortunately it is non-trivial to override the tagged template literals. They exist roughly in the following form:

```tsx
interface TemplateResult {
    strings: string[] // array of the literal strings in the template
    values: unknown[] // array of the values to interpolate between the strings. There will be one less of these
}
```

and modifying these will be both non-trivial and extremely prone to breakage

## Rough Approach 

In this PR, I've introduced a `lit-mangler` for mangling the templates at build time. It works roughly like this:
1. Parse the Typescript syntax tree for upstream's `.html.ts` file
2. Extract all the `html` tagged template literals into a tree (as literal can exist inside literals for ternaries/maps/children ect).
3. Create a mangler script for the upstream template
4. Select one of these literals (i.e. by contains `id="my-id"`)
5. Load the HTML of the template into a `DocumentFragment`
6. Let the mangler script modify the fragment using DOM methods (this should make migrating our Polymer templates straightforward).
7. Reserialize the `DocumentFragments` to the original script and write them to the `preprocess` directory

## Alternatives Considered

See https://github.com/brave/brave-core/pull/27661 for an alternative runtime approach. I think the solution in this PR is much more robust and should also be more performant, as it won't thrash Lit's template cache every render.

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Demonstration

As part of this PR, I've rewritten an existing override which broke when upstream migrated their element from Polymer to Lit. I can remove that from this PR, if it makes things easier.

## Rough Example:

```ts
// in brave/chromium_src/chrome/browser/resources/extensions/detail_view.mangle.html.ts
import { mangle } from 'lit_mangler'

// Note: This will error if we can't find the template to mangle
mangle((element) => {
    const section = element.querySelector('.section-content')
    if (!section) {
        throw new Error('[Brave Extensions Overrides] Could not find .section-content. Has the ID changed?')
    }
    section.textContent = '$i18n{privateInfoWarning}'
    section.append('<span ?hidden=${!this.data.incognitoAccess.isActive}> $i18n{spanningInfoWarning}</span>')
    section.append('<span> $i18n{privateAndTorInfoWarning}</span>')
}, t => t.text.includes('id="allow-incognito"'))
```

which replaces the old, broken override brave/chromium_src/chrome/browser/resources/extensions/detail_view.ts
```ts
import {RegisterPolymerTemplateModifications} from 'chrome://resources/brave/polymer_overriding.js'
import {loadTimeData} from 'chrome://resources/js/load_time_data.js';

RegisterPolymerTemplateModifications({
  'extensions-detail-view': (templateContent: DocumentFragment) => {
    let optionsTemplate: HTMLTemplateElement | null =
      templateContent.querySelector<HTMLTemplateElement>('template[is="dom-if"][if*="shouldShowOptionsSection_"]')
    if (!optionsTemplate) {
      console.error('[Brave Extensions Overrides] Could not find optionsTemplate')
      return
    }
    let incognitoTemplate =
      optionsTemplate.content.querySelector<HTMLTemplateElement>('template[is="dom-if"][if*="shouldShowIncognitoOption_"]')
    if (!incognitoTemplate) {
      console.error('[Brave Extensions Overrides] Could not find incognitoTemplate')
      return
    }
    let incognitoWarningDiv = incognitoTemplate.content.querySelector<HTMLDivElement>('.section-content')
    if (!incognitoWarningDiv) {
      console.error('[Brave Extensions Overrides] Could not find incognitoWarningDiv')
      return
    }
    incognitoWarningDiv.innerText = loadTimeData.getString('privateInfoWarning')
    const spanningWarningSpan = document.createElement('span')
    spanningWarningSpan.setAttribute('class', 'section-content')
    spanningWarningSpan.setAttribute('hidden', '[[data.isSplitMode]]')
    spanningWarningSpan.innerText =
        ' ' + loadTimeData.getString('spanningInfoWarning')
    const privateAndTorWarningSpan = document.createElement('span')
    privateAndTorWarningSpan.setAttribute('class', 'section-content')
    privateAndTorWarningSpan.innerText =
        ' ' + loadTimeData.getString('privateAndTorInfoWarning')
    incognitoWarningDiv.appendChild(spanningWarningSpan)
    incognitoWarningDiv.appendChild(privateAndTorWarningSpan)
  }
})

export * from './detail_view-chromium.js'
```

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

fallaciousreasoning commented on chromium_src/tools/grit/preprocess_if_expr.py:
```diff
@@ -43,6 +43,14 @@ def maybe_keep_upstream_version(override_in_folder, out_folder, override_file):
+
+    # Note: We read from and write to the preprocess file - this way any
+    # preprocessing that upstream does will be mangled.
+    subprocess.run(['npx', 'tsx', '--tsconfig', '../../brave/tsconfig.json', '../../brave/build/commands/scripts/lit_mangler.ts', 'mangle', '-m', mangler_file, '-i', preprocess_file, '-o', preprocess_file])
```
Hey @thypon this seems like it'd be a good candidate for a `semgrep` rule - you shouldn't be able to execute `npx` without the `--no` flag

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/lit-mangling-static/storybook.html)

diracdeltas commented on build/commands/scripts/lit_mangler.ts:
```diff
@@ -0,0 +1,200 @@
+    const world = new JSDOM()
+    const document = world.window.document
+    const element = document.createElement('template')
+    element.innerHTML = template.text
```
@thypon can you review this please

thypon replied:
This is server side, and build time, so looks fine.
It is also not "executed" as a site, since in `JSDOM`.

Are we sure we want to use `JSDOM` here and not a chromium-headless to interpret the DOM?
These two DOMs are not equivalent.

fallaciousreasoning replied:
I think this is okay, as per https://github.com/brave/reviews/issues/1872#issuecomment-2670119360

diracdeltas commented on package.json:
```diff
@@ -171,6 +171,7 @@
     "timers-browserify": "2.0.12",
     "ts-jest": "28.0.8",
     "ts-loader": "9.5.2",
+    "tsx": "4.19.2",
```
as mentioned in sec review, prefer to get rid of this

goodov commented:
Okay, the approach is interesting, but we can improve things here:

1. The placement of the script is a bit questionable. This is not a `npm run ...` command we usually put into `build/commands`. In my opinion the mangler should go into `//brave/tools/chromium_src/lit_mangler`.

1. Consider splitting the mangler into a library and cli files to remove the awkward process.argv check.

1. (most important) You wired things to run this TS tool without any type checking during all (including Release) browser builds. This is not cool, we need to have type checking in some form here. I see few possible options:
    * Figure a way to run `tsc --noEmit` on the tool sources on CI.
    * Figure a way to compile `.ts` into `.js` during the build and call `.js` version directly. It may be possible to have this as a GN build dependency to make sure `.js` is generated before any `preprocess_if_expr.py` tries to use it.

petemill commented on build/commands/scripts/lit_mangler.ts:
```diff
@@ -0,0 +1,200 @@
+    return root
+}
+
+const injectPlaceholders = (template: HTMLTemplateTags) => {
```
A couple of these internal functions might be a good place to add comments so a reader can understand how this works, e.g. why we need to add placeholders

petemill commented on chromium_src/tools/grit/preprocess_if_expr.py:
```diff
@@ -44,17 +44,42 @@ def maybe_keep_upstream_version(override_in_folder, out_folder, override_file):
     return None
 
 
+def run_mangler(mangler_file, preprocess_file):
+    """Runs the mangler on the given file"""
+    import subprocess
+
+    ts_config = os.path.join(get_src_dir(), 'brave', 'tsconfig.json')
+    lit_mangler = os.path.join(get_src_dir(), 'brave', 'build', 'commands',
+                               'scripts', 'lit_mangler.ts')
+
+    # Note: We read from and write to the preprocess file - this way any
+    # preprocessing that upstream does will be mangled.
+    subprocess.run([
+        'npx', '--no', '--', 'tsx', '--tsconfig', ts_config, lit_mangler,
+        'mangle', '-m', mangler_file, '-i', preprocess_file, '-o',
+        preprocess_file
+    ],
+                   check=True)
+
```
Seems like injecting the mangle script run after preprocess is the simplest and most universal place. And if we want to type check we could run `tsc` here. Even though it's going to run separately for each file, it should be fast?
But it seems like we _could_ do this at a higher level as part of the `build_webui` GN template after the preprocess stage, since all lit .html.ts files will run through that target. Not sure if doing it in GN will help with any issues that come up in review.
Also, if we're having any performance issues perhaps we could run against all the files in the directory or whatever list `preprocess_if_expr` has. But you'd need to turn the module into a class or similar.

fallaciousreasoning replied:
Yeah, I'm not sure what the best solution here is - at least with this, it will only be rerun when the target the manglers exist in change, whereas if this was part of `build_webui` it would have to run whenever any of the manglers changed, which I'm not sure would be better.

Additionally, it would make ordering quite difficult as this needs to run after the Chromium files are preprocessed (so it can run over the preprocessed files and doesn't get overwritten) but before the Typescript build happens.

petemill reviewed (commented):
It's exciting to think about doing this at compile time instead of run time. For modifying the lit component itself, we'd still use a combination of src_overrides and `lit_overriding.ts` - though perhaps the styles could be injected using this technique instead?

Using tsx seems ok because we don't force a bundled node version, so it just uses local node script same as our brave_node.py or `npm run ...` does.

thypon commented on build/commands/scripts/lit_mangler.ts:
```diff
@@ -0,0 +1,200 @@
+
+    mangler(element.content)
+
+    template.text = element.innerHTML
```
this does not look right

fallaciousreasoning replied:
I think it is because there's an encode which happens when parsing to HTML and we want things to roundtrip nicely:

```
// the string
"<div>${"<foo&"}</div>"

// is parsed to
<div>${&quot;&lt;foo&amp;&quot}</div>

// which should roundtrip back to the string
"<div>${"<foo&"}</div>"

```

This is because its valid to have `html` which is really just a string inside the interpolated content (in the real world it gets replaced with the result of parsing to JS). However, we want the raw string from inside the interpolation so we can edit it.

fallaciousreasoning commented:
> 1. The placement of the script is a bit questionable. This is not a npm run ... command we usually put into build/commands. In my opinion the mangler should go into //brave/tools/chromium_src/lit_mangler.

That's a much better location. Moved it!

> 2. Consider splitting the mangler into a library and cli files to remove the awkward process.argv check.

I actually did this originally but moved them into one :laughing: Split them again.

> 3. (most important) You wired things to run this TS tool without any type checking during all (including Release) browser builds. This is not cool, we need to have type checking in some form here. I see few possible options

I've gone with a two pronged approach:
1. Run `tsc --noEmit` on each mangler when we run it. This unfortunately involves creating a temporary tsconfig file because `tsc` doesn't support path maps as command line args.
2. Run `tsc -p tsconfig-mangle.json` as part of presubmit (which ensures `lit_mangler_cli` is checked).

> 4. npx should be avoided during the build as it's a pretty slow thing. Pls use something like brave_node.py to execute the required .js file directly.

Calling via `brave_node.py` now :smile:

fallaciousreasoning commented on tools/chromium_src/lit_mangler.ts:
```diff
@@ -0,0 +1,281 @@
+                .replaceAll('"', '&quot;')
+                .replaceAll('<', '&lt;')
+                .replaceAll('>', '&gt;')
+        })
```
**Note:** This is at build time

thypon replied:
this might still introduce issues, since we might demangle a safety mangled element

fallaciousreasoning replied:
I think it should be safe because `&` is the first thing we encode so all HTML encoded entities should be converted and `&amp;` is the last thing we decode.

`&lt;script&gt;alert('hahaha')&lt;script&gt;` ==> `&amp;lt;script&amp;gt;alert('haha')&amp;lt;script&amp;gt;` ==> `&lt;script&gt;alert('hahaha')&lt;script&gt;`

```ts
// in the upstream file:
html`<div>Hi from Jay &amp; Andrea</div>`

// mangled
<div>Hi from Jay &amp;amp; Andrea</div>`
// we replaced all `&`s with &amp;

// output
html`<div>Hi from Jay &amp; Andrea</div>`
// we replaced all `&amp;`s with `&`

thypon replied:
Can we add some test re mangling/demangling here?

thypon commented:
@fallaciousreasoning the demangling looks still somewhat problematic

petemill commented:
I suggested to @fallaciousreasoning that we could alleviate any concerns about unintentional html modifications by automatically generating snapshot files for each template that we're modifying. A unit test could generate a .patch file representing the change from upstream's original `.html.ts` file, and this could be committed. The unit test could fail if the file hasn't been committed - in CI the test would fail if the generated patch differs from the existing local patch, whereas locally the test would write the new version of the patch file to disk. This could make code review for any modifications easier and give everyone confidence that JSDom doesn't destroy information.

It also solves one of the issues working with the existing polymer template-modification system (and is still an issue with this new lit version) - it's difficult to know if your modifications were applied successfully.

If it's too much of a challenge we could instead just have a couple simple tests that verify JSDom doesn't remove attributes we care about.

fallaciousreasoning commented:
@petemill I've added the mangler snapshots - at the moment its a diff between the upstream file and our mangled file, but I could as easily add the whole mangled file as the snapshot (which might make reviewing diffs easier).

The test will fail when the snapshots don't match and we can easily update them with Jest :smile:

goodov commented on tools/chromium_src/lit_mangler/lit_mangler_cli.ts:
```diff
@@ -0,0 +1,76 @@
+    tsConfig.compilerOptions.baseUrl = baseDir
+
+    // Write the tsconfig to a temporary file.
+    const tempDir = os.tmpdir()
```
please do not use `tmpdir`. The build should be contained in `out/` directory.

goodov commented on tools/chromium_src/lit_mangler/__snapshots__/mangle.test.ts.snap:
```diff
@@ -0,0 +1,25 @@
+exports[`mangled files should have up to date snapshots ./chromium_src/chrome/browser/resources/extensions/detail_view.mangle.html.ts should match snapshot 1`] = `
+"===================================================================
+--- ../chrome/browser/resources/extensions/detail_view.html.ts
++++ ../out/Component/gen/chrome/browser/resources/extensions/preprocessed/detail_view.html.ts
```
don't think this will work. we have a lot of other bulid configurations.

fallaciousreasoning replied:
Updated it to skip the outdirectory in the path - it will now use `gen/....`

goodov replied:
the snapshot still says `../out/Component/gen`. Assuming you just didn't push the fixed version.

fallaciousreasoning replied:
Yeah didn't push the fixes, as I got halfway through running from the out directory and realised I'd need to bundle the files

goodov commented on chromium_src/tools/grit/preprocess_if_expr.py:
```diff
@@ -92,6 +117,13 @@ def parse_args(self, original_method, argv):
+                                                      '.html.ts')
+                run_mangler(os.path.join(override_root_folder, override_file),
+                            os.path.join(out_folder, upstream_file))
+                continue
```
do we expect both overrides to exist for a file (mangle and not mangle)? Does it make sense? If not, let's drop `.mangle` extension and remove `chromium_src` customization to support `.mangle` files. Instead to determine if a mangler should be run, scan the `.ts` file for `from 'lit_mangler'`.

fallaciousreasoning replied:
Happy to do that but the reason I introduced the `.mangle` extension was to make it obvious it's doing something different (i.e. at build time, rather than at runtime).

goodov replied:
That's fine, I agree it would be better to differentiate those files. Let's generalize the chromium_src overrides lookup a bit (see the comment in `util.js`).

Maybe make those files be more explicit with the extension? Say, `.lit_mangle.ts`?

fallaciousreasoning replied:
Yup, sounds good to me :smile:

goodov commented on tools/chromium_src/lit_mangler/lit_mangler_cli.ts:
```diff
@@ -0,0 +1,76 @@
+}
+
+const runTypecheck = (files: string[]) => {
+    const result = childProcess.spawnSync('tsc', ['-p', getTsConfigForFiles(files)])
```
pretty sure this will generate `.js` files in the output directory. You either use those or pass `--noEmit`. Pls try to use the generated `.js` directly and drop `tsx` from the build process entirely.

You can use `gen/<directory_determined_by_input_file>/tsconfig.json` and put generated files there.

`tsx` is good for single-use scripts, but here we'd better to have a more strict process.

fallaciousreasoning replied:
`noEmit` is in the `tsconfig.json` so it won't be generated.

fallaciousreasoning replied:
So if we want to run the generated files we're going to need a build/bundle step, as well as a compile step.

This would be to support:
1. Path mapping (i.e. `import { mangle } from 'lit_mangler'`)
2. node_modules (i.e. `import ts from 'typescript'`, which doesn't work because the generated files won't be in the `brave` directory where node_modules live).

This isn't insurmountable but its going to make the whole build process more complicated for what I think is relatively little gain. FWIW, all our other build scripts are in Javascript, not Typescript. The output of these scripts is Typescript which will be typechecked as a normal part of the build.

I agree that we should be typechecking the build scripts but I'm not convinced the way to do that is to bundle them into Javascript in our build directory.

If the concern is another dependency (`tsx`) and build time performance I could look into using `--experimental-strip-types` (https://nodejs.org/en/learn/typescript/run-natively#running-typescript-code-with-nodejs) which is supported in the latest LTS version of NodeJS instead of `tsx`.

Using `tsx` is nice because it means we don't need to worry about aligning `node` aliases and `tsconfig` paths, and personally I think it makes things a lot less complex here. I'm willing to be convinced, but I'd prefer this work the same way we'd like it to work for `build/commands/scripts/commands.js`, in a hypothetical world where we migrate that to Typescript.

goodov replied:
> `noEmit` is in the `tsconfig-mangle.json` so it won't be generated. I'll try and migrate

My bad, sorry, you're right.

> Using `tsx` is nice because it means we don't need to worry about aligning `node` aliases and `tsconfig` paths, and personally I think it makes things a lot less complex here. I'm willing to be convinced, but I'd prefer this work the same way we'd like it to work for `build/commands/scripts/commands.js`, in a hypothetical world where we migrate that to Typescript.

Re the overall approach: I agree that this might be a not so easy task to implement this as a `tsc` call followed by node execution of `js` files in a clear and easy way.

I guess we can stick with `tsx`, but in case of any issues with it I'd expect you to jump in and debug and solve the issues on your own :) 

Another possible approach I can think of here is to have a single `tsc` call somewhere (maybe a GN target) to compile all `*.mangle.ts` files from `chromium_src` dir along with the mangler itself (the way you listed it in `tsconfig-mangle.json`), and then call the transpiled `js` files where we need to.

fallaciousreasoning replied:
> Another possible approach I can think of here is to have a single tsc call somewhere

That still has the problem that we'd need to bundle in `node_modules` (or override `NODE_PATH`) and setup some bundler to work with the path aliases.

> I guess we can stick with tsx, but in case of any issues with it I'd expect you to jump in and debug and solve the issues on your own :)

If it's okay with you, I'd be keen to trial `tsx` for this and see if any problems come up (definitely happy to own any issues). If we start encountering problems I'll set up bundling but I'd really prefer to avoid it as I think it makes the build process a lot more complicated.

fallaciousreasoning commented on PRESUBMIT.py:
```diff
@@ -49,6 +49,36 @@ def _web_files_filter(affected_file):
 
+
+# Ensure Typescript files which we use as part of the build process are typechecked.
+def CheckTypescriptBuildFilesCompile(input_api, output_api):
```
Okay, I'll try and remove all the `tsx` related stuff - I had an idea that it'd be nice to try and migrate all our build scripts to Typescript.

goodov commented on build/commands/lib/util.js:
```diff
@@ -551,6 +551,13 @@ const util = {
+
+      // .mangle.html.ts files are used to modify the upstream .html.ts file at
+      // build time.
+      if (overriddenFile.endsWith('.mangle.html.ts')) {
```
let's make this in a more generic way where we list all possible override extensions.

```javascript
      const overrideExtensions = [
        '.mangle.ts',
      ]

      for (const ext of overrideExtensions) {
        if (overriddenFile.endsWith(ext)) {
          overriddenFile = overriddenFile.slice(0, -ext.length)
          break
        }
      }
```

pls make sure to do the same for python chromium_src checker.

fallaciousreasoning commented on build/commands/lib/util.js:
```diff
@@ -551,6 +551,18 @@ const util = {
+      const additionalExtensions = {
+        // .lit_mangler.html.ts files are used to modify the upstream .html.ts file at
+        // build time.
+        '.lit_mangler.html.ts': '.html.ts',
```
Updated - now we just append the `.lit_mangler.ts` to then upstream path like you suggested :smile:

![image](https://github.com/user-attachments/assets/04a52643-8181-4cea-88fa-990da3091375)

fallaciousreasoning commented:
**Note:** The audit failure on CI is from the rust `audit_deps`

fallaciousreasoning commented on build/commands/lib/util.js:
```diff
@@ -551,6 +551,18 @@ const util = {
       const relativeChromiumSrcFile = chromiumSrcFile.slice(chromiumSrcDirLen)
       let overriddenFile = path.join(config.srcDir, relativeChromiumSrcFile)
+
+      const additionalExtensions = [
```
These are additional extensions which can be used to override the upstream file.

https://github.com/brave/brave-core/pull/27703#discussion_r1968841447

`foo.html.ts` can be overridden by `foo.html.ts.lit_mangler.ts`

We're going with a different extension here to make it clear that this does something different (i.e. it runs at build time, rather than at runtime).

bridiver replied:
this is a bit confusing imo, why are the extensions different in the first place?

fallaciousreasoning replied:
> this is a bit confusing imo, why are the extensions different in the first place?

The extension is different to make it clear that this doesn't work the same way our other overrides work. This is much more similar to Plaster than to most of our other `chromium_src` overrides. Its a way of modifying the upstream source in a structured way.

> I would really prefer not to be copying files like this in the first place in general. I guess we are already doing, but we should be trying to minimize it.

AIUI this isn't copying files - it touches the upstream file when something in our codebase that could change it changes. Still not ideal but I think its worth it for the ergonomics.

goodov commented on chromium_src/tools/grit/preprocess_if_expr.py:
```diff
@@ -44,15 +44,44 @@ def maybe_keep_upstream_version(override_in_folder, out_folder, override_file):
+    import brave_node
+
+    tsx = brave_node.PathInNodeModules('tsx', 'dist', 'cli.mjs')
+    ts_config = os.path.join(get_src_dir(), 'brave', 'tsconfig-mangle.json')
```
nit: `wspath('//brave/tsconfig-mangle.json')`. same for the path below.

fallaciousreasoning replied:
Great tip :+1:

goodov reviewed (approved):
This approach is interesting, I'm fine with trying it out to see how well it behaves in the real world.

Although there are some questionable html manipulations are going on, but we're basically editing CST by hand to some degree here. As long as we validate the correctness of the produced output (the build process runs `tsc` on mangled `.ts` files after all), we should be fine.

petemill commented on tools/chromium_src/lit_mangler/__snapshots__/mangle.test.ts.snap:
```diff
@@ -0,0 +1,25 @@
+"===================================================================
+--- ../chrome/browser/resources/extensions/detail_view.html.ts
++++ gen/chrome/browser/resources/extensions/preprocessed/detail_view.html.ts
+@@ -342,14 +342,12 @@
```
I think the purpose is actually exactly to _help_ the rebase process. When the upstream files change in a rebase, the rebase team can simply cause the snapshots to all be updated, and then we get to review in the rebase PR that the existing template modifications are still applying. What do you think about that?

bridiver replied:
cc @mkarolin @cdesouza-chromium

fallaciousreasoning replied:
I had a chat to @cdesouza-chromium on Slack. He's okay to see how this goes.

fallaciousreasoning commented on tools/chromium_src/lit_mangler/lit_mangler.ts:
```diff
@@ -0,0 +1,284 @@
+                .replaceAll('&lt;', '<')
+                .replaceAll('&quot;', '"')
+                .replaceAll('&amp;', '&') // Note: &amp; needs to be last
+        })
```
Hmmm, I don't think kind of mode would make sense. How could you tell what content is text and what content is HTML?

Consider:
```html
<!-- This is not the same -->
<div>${"&lt;script&gt;alert('hello')&lt;/script&gt;"}</div>
<!-- as this -->
<div>${"<script>alert('hello')</script>"}</div>
```

the reason we can replace the HTML entities is because we're serializing to a JavaScript tagged template literal, not HTML, which JSDom is dealing with.

fallaciousreasoning commented on PRESUBMIT.py:
```diff
@@ -147,7 +177,11 @@ def CheckESLint(input_api, output_api):
+
+    # ESLint struggles with the dynamic import for the mangler here, and it
+    # isn't possible to disable that specific check.
+    files_to_skip = input_api.DEFAULT_FILES_TO_SKIP + (
```
:facepalm: I actually knew about this, I dunno what I was thinking. Thanks!

fallaciousreasoning commented:
Waiting on https://github.com/brave/devops/pull/13328 for cross compiling unittests

goodov commented:
@fallaciousreasoning I keep seeing how you update this PR :). Do you wait for something? To me it seems good to merge.

fallaciousreasoning commented:
@goodov there were a couple of things:
1. I was on PTO for a week and a bit, and wanted to make sure I was fully available before merging
2. Windows tests were failing because the diff in the test was looking like Windows (with `\r\n` and `\\` instead of `/` in paths. 

I've fixed the second and I'm back, so I'm going to merge :smile: Finished things off last night (had to get my Windows VM working again).

brave-builds commented:
Released in [v1.79.30](https://github.com/brave/brave-browser/releases/tag/v1.79.30)

bridiver commented on build/commands/lib/guessConfig.ts:
```diff
@@ -0,0 +1,60 @@
+
+// This uses environment variables as there is currently no way to pass custom
+// arguments to the |storybook build| cli.
+config.update({
```
you also don't have a complete list of build config directories here because there are variations on these and I would not rely on `most recently built` in any case
